### ISSUE TITLE
Upgrade with an unknown signer doesn't update deployment

### ISSUE THREAD

**Issue by gisek** [wighawag/hardhat-deploy#259]
**Describe the bug**
`deployments/network/ContractName.json` file is not updated when an upgrade with an unknown signer is performed

**To Reproduce**
1. Deploy a contract as `OptimizedTransparentProxy`
2. Create a new version of the contract `ContractNameV2` and add a new function `newFunction()`
3. Deploy upgrade like this:
```
  await hre.deployments.catchUnknownSigner(
    hre.deployments.deploy("ContractName", {
      from: deployerAddress,
      proxy: {
        proxyContract: "OptimizedTransparentProxy",
        owner: newProxyAdminOwnerAddress, // make sure this one is unknown to hardhat!
      },
      contract: "ContractNameV2",
      args: []
    })
  );
```
4. Execute the instruction form console (replace the implementation in a proxy)
5. In a hardhat task try to call the new function: 
```
const token = await hre.ethers.getContract("ContractName") as ContractNameV2;
const newFieldValue = await token.newFunction();
```
6. You get `TypeError: token.newFunction is not a function`
7. It works fine if you do it like this:
```
const token = await hre.ethers.getContractAt("ContractNameV2", proxy_address) as ContractNameV2;
```

**Expected behavior**
`deployments/network/ContractName.json`  Should be updated and `await hre.ethers.getContract("ContractName")` should return the right ABI

**versions**
 - hardhat-deploy 0.9.27
 - hardhat  2.6.8
 - nodejs v14.17.5

**Additional context**

1. If you use a known signer there it works jus fine
2. It look like this line throws: https://github.com/wighawag/hardhat-deploy/blob/19b4ef8f9b1423ade4c318178d356f1747cadf63/src/helpers.ts#L1520
3. and therefore this line never gets called: https://github.com/wighawag/hardhat-deploy/blob/19b4ef8f9b1423ade4c318178d356f1747cadf63/src/helpers.ts#L1564

**Comment by wighawag** [comment]
Before step 5, you ll need to reexecute the deployment scripts so that it update the deployments file

**Comment by AndrzejGisWideLab** [comment]
@wighawag The `deployments/network/ContractName_Implementation.json` gets updated right away `deployments/network/ContractName.json` doesn't get updated even if I rerun the deployment task.

What is the recommended way to do this? If I just run the deployment task again it changes nothing.

**Comment by wighawag** [comment]
> The deployments/network/ContractName_Implementation.json gets updated right away deployments/network/ContractName.json doesn't get updated even if I rerun the deployment task.

this is expected since it deploy that

but you need to rexecute the deploy script to update the link with the proxy. if that link is not updated, there is probably an issue. 

when you reexecute the scripts before step 5, what happen, are your scripts idempotent ? do you get any error, are you told to perform a new upgrade ?

**Comment by AndrzejGisWideLab** [comment]
@wighawag Yes the scripts are idempotent thanks to the framework. The deployment framework (I thought it's actually the `hardhat-deploy` itself that does it) appends the script name to the `deployments/network/.migrations.json` and the script never executed again. This is why I'm wondering about the procedure here.

this code:
```
  await hre.deployments.catchUnknownSigner(
    hre.deployments.deploy("ContractName", {
      from: deployerAddress,
      proxy: {
        proxyContract: "OptimizedTransparentProxy",
        owner: newProxyAdminOwnerAddress, // make sure this one is unknown to hardhat!
      },
      contract: "ContractNameV2",
      args: []
    })
  );
```

will be executed once only, since after the first successful run it's file will be appended to `deployments/network/.migrations.json`. Duplicating this code in a subsequent script/migration doesn't seem right. 

Regarding you second question, the step 4 is the last deployment step that I perform. Step 5 is just a simple hardhat task which I'm running afterwards with intention to test it the upgrade worked.

**Comment by wighawag** [comment]
Ha! I see. I did not though about someone using the migration option. 

Note that if you have your script idempotent, you do not need to use the migration options though.

If you want to keep using it, I am wondering what hardhat-deploy could do to make it easier for you to update the deployment file.

One thing that it possible in the same file it is to catchUnknowSigner but then pause the script and then when you are done with the upgrade, the scripts perform the upgrade again but this time as it will not change, it will update the deployment

**Comment by AndrzejGisWideLab** [comment]
@wighawag Thanks! I'll give it a go

**Comment by AndrzejGisWideLab** [comment]
@wighawag It worked. However I think it would be better if the `deployments/network/ContractName.json` file was updated right away along with proxy and implementation. Maybe I'm not seeing the whole picture here, but I can see no reason to run the script twice or to add another migration with basically the same code. When `deployments/network/ContractName.json` is out of sync with proxy and implementation, the code is in a kind of corrupt state imho.

**Comment by wighawag** [comment]
@AndrzejGisWideLab how would that work ?
when catchUnknownSIgner catch the error, this is when you execute your upgrade, at that point, you need to tell hardhat-deploy to update. there is no background process running watching the chain for updates.

I guess you could create a utility function that call `catchUnknowSIgner` and keep watching the chain and then update. Not sure if it need to be part of hardhat-deploy though.

**Comment by AndrzejGisWideLab** [comment]
@wighawag I didn't dive very deep into the code but basically what I think is:
Whatever updates these 2 json files (it happens **before** the the implementation is replaces on-chain):
`deployments/network/ContractName_Proxy.json`
`deployments/network/ContractName_Implementation.json`
Could immediately update also:
`deployments/network/ContractName.json`

I don't understand what's the purpose of splitting those .json files updates into 2 stages, but like I said, I might not be seeing the big picture here.

**Comment by wighawag** [comment]
I could potentially add a command `update-proxies` that would check for new implementation and update accordingly

**Comment by AndrzejGisWideLab** [comment]
> I could potentially add a command `update-proxies` that would check for new implementation and update accordingly

This would be great!

**Comment by masha256** [comment]
Hey all, I recently came across this exact issue. In our case, we use `catchUnknownSigner` and then save the output to then schedule an update with our multi-sig.

In my opinion, the core issue is that its fundamentally a two-stage process:

1) Get the new implementation deployed (and possibly tested)
2) Update the ABI of the proxy to know about any changes now available in the updated version

So I do agree that `ContractName.json` should not automatically be overwritten at new implementation deploy time, since, by definition, that new ABI is not available yet until (in our case) the multi-sig does the upgrade.

So yeah, a new command we can run after the multi-sig `upgrade` call, that updates the the main `ContractName.json` would be ideal, as we want our repo ABI to represent whats 'current' on chain.

As of now, we have been grabbing the new ABI out of the updated `ContractName_Implementation.json` and pasting it into the ContractName.json ABI after the on-chain `upgrade` is performed. Obviously not ideal, and ripe for a copy/paste error.

**Comment by Archethect** [comment]
We have the same problem on our side and I'm thinking about what the best solution would be to solve this in a clean way.

**Comment by wighawag** [comment]
I am currently working on hardhat-deploy v2 and the same issue will be present there and i think adding a new command that "sync" the deployment files make sense

It can read the implementation address onchain and check for the matching implementation deployment file and update the abi,etc..
